adr/ADR-15.md

```diff
@@ -0,0 +1,231 @@
+
+If the result of this request is "OK", it means that the server has created the JS consumer and uses the deliver subject the NATS subscription is created on. At this point, the jetstream subscription will keep track of the fact that it did create the JetStream consumer and will delete it on Unsubscribe() or Drain(). For ephemerals, the consumer name will be saved from the creation request's response, since it was not available beforehand.
+
+If the result indicates that the consumer already exists, then it means that there was a race where a process got a lookup "not found" error, but when attempting to create the JetStream consumer, got an "already exists" error. In this case, the library will perform a consumer lookup and for branch to the description above that describes what to do when getting a consumer info. Note that for pull subscriptions, basic checks of consumer type validity should be done, but not the checks specific to "push" consumers. For them, unless this is a queue subscription, the API call will return an error.
+
+When the consumer already exists, for push consumers, the NATS queue subscription that was created prior to the "AddConsumer" call needs to be destroyed and replaced with the new NATS queue subscription on the consumer info's `DeliverSubject`.
+
+### Runtime
+
+#### Heartbeats
+
+If the JetStream consumer is configured with heartbeats, the server will periodically (based on the specified heartbeat interval) send hearbeats containing meta information about the stream and consumer sequences.
+
+The library should setup a timer to monitor that heartbeats from the server are properly received. No specific action is taken by the library other than notifying the user through the asynchronous error callback if heartbeats are detected missing.
+
+Also, the library should always check for JetStream heartbeat (and flow control) status messages, regardless of how the subscription was created. That is, the checks should not be conditional to the `IdleHeartbeat` option passed to a subscribe API call.
+
+The library should keep track of all messages which reply subject start with `$JS.ACK` since they contain information about the stream and consumer sequence, etc..
+
+When the library receives a message that is found to be a control message (no payload and containst the "Status" header), if the header is "100 Idle Heartbeat", then the library should check if there is any gap detected between what was the last tracked message meta information and the content of the heartbeat. If a gap is detected, an asynchronous error should be reported indicating what the gap is.
+
+Ordered consumers (should have its own ADR) will handle all that for the application, recreating the subscription as needed at the right stream sequence.
+
+**Note**: In most libraries the hearbeats messages are never presented to the user. Those are handled internally.
+
+#### JS Ack
```

Comment:
It would be more clear if we call this the `DeliveryInfo` - this is the data encoded into the reply subject of a JetStream message sent to a subscriber.

Reply:
Added a bit of verbiage to clarifies that we are talking about the ACK/delivery information for a message sent by the server to a subscriber.